src/main/java/org/jline/utils/AttributedStyle.java

```diff
@@ -42,14 +36,22 @@
     static final int F_CROSSED_OUT  = 0x00000080;
     static final int F_FOREGROUND   = 0x00000100;
     static final int F_BACKGROUND   = 0x00000200;
+    static final int F_GENERIC_ESCAPE  = 0x00400;
 
-    static final int MASK           = 0x000003FF;
+    static final int MASK           = 0x000007FF;
 
     static final int FG_COLOR_EXP    = 16;
     static final int BG_COLOR_EXP    = 24;
     static final int FG_COLOR        = 0xFF << FG_COLOR_EXP;
     static final int BG_COLOR        = 0xFF << BG_COLOR_EXP;
 
+    public static final AttributedStyle DEFAULT = new AttributedStyle();
+    public static final AttributedStyle BOLD = DEFAULT.bold();
+    public static final AttributedStyle BOLD_OFF = DEFAULT.boldOff();
+    public static final AttributedStyle INVERSE = DEFAULT.inverse();
+    public static final AttributedStyle INVERSE_OFF = DEFAULT.inverseOff();
+    public static final AttributedStyle GENERIC_ESCAPE =
+        new AttributedStyle(F_GENERIC_ESCAPE, F_GENERIC_ESCAPE);
 
```

Comment:
The problem I see is that your solution is somewhat mixing structural markup and styling, as it's resetting the styles when exiting the markup afaik. 
What about going back to my initial suggestion : i.e. a simple HIDDEN style that can be turned on/off and which turns characters to 0 length in computation.  It should not affect split / substring, given it's 0 length.  It would behave as the other style attributes and is expected not to affect other styles.

There's a problem in expecting the generic escape to reset the style imho, as all the styling operations that happen on the AttributedString are based on flag changes, not the current state.  For example, you can't have an single AttributedString output ```{turn on bold}foo{turn off bold}{turn on bold}bar{turn off bold}```.  If we really need to handle such thing, we could add an additional flag that forces a style reset before or after the character, but that's a bit more tricky.  But in your case, you can simply set back the style to DEFAULT before starting a HIDDEN sequence if you know the sequence will affect the styling.